### DISCUSSION TITLE
[Transit nodes] Confuse about the parallelism when using transit nodes

### DISCUSSION THREAD

**Discussion by motss** [vercel/turborepo#8387] [Help] [answered]
### Summary

> If you test out making breaking changes in your ui package again, you'll notice that the caching behavior is now correct. However, tasks are no longer running in parallel.

I am reading the new [docs about transit nodes](https://turbo.build/repo/docs/crafting-your-repository/configuring-tasks#dependent-tasks-that-can-be-ran-in-parallel) and I don't get it why and how the parallelism works for the case with transit nodes but not with the case without transit nodes.

Please help validate if my graphs below are correct in terms of the execution order with and without transit nodes. Thanks. 🙏 

```sh
# Note: `dependsOn` in the `Dependencies` has nothing to with `dependsOn` field in `turbo.json`.
# It basically means `a` installed `b` as a dependency package for `a -- dependOn --> `b.

# Example 1
# Dependencies
# a -- dependsOn --> b
# b -- dependsOn --> c

# --- without transit nodes ---
c#build
       b#build
              a#build

# --- with transit nodes ---
a#transit
                       a#build
b#transit
                b#build
c#transit
         c#build
```

```sh
# Example 2
# Dependencies: 
# a & b -- dependsOn --> c

# --- without transit nodes ---

b#build
       a#build
       c#build

# --- with transit nodes ---

a#transit
                a#build
b#transit
         b#build
c#transit
                c#build
```

### Additional information

```js
N/A for the discussion stated above.
```

### Example

N/A

**Comment by anthonyshew** [comment]
Can you clarify what those examples are supposed to mean? It's impossible to tell what you meant.

**Reply by motss** [reply]
The examples are the charts drawn to outline the execution order of the `turbo` tasks which inspired by this [page](https://turbo.build/repo/docs/crafting-your-repository/configuring-tasks) in the official doc, refer to the images below:

<img width="2056" alt="serial execution" src="https://turbo.build/_next/image?url=%2Fimages%2Fdocs%2Fslow-tasks-dark.png&w=1920&q=75&dpl=dpl_3VFYvdEA5y3sv4voiVPtpXf6vuDf">

<img width="2056" alt="parallel execution" src="https://turbo.build/_next/image?url=%2Fimages%2Fdocs%2Fturborepo-tasks-fast-dark.png&w=1920&q=75&dpl=dpl_3VFYvdEA5y3sv4voiVPtpXf6vuDf">

**Reply by anthonyshew** [reply, marked as answer]
The best explanation we're likely able to give is probably from this documentation: https://turbo.build/repo/docs/crafting-your-repository/configuring-tasks#dependent-tasks-that-can-be-ran-in-parallel

**Reply by rtarpine-vanta** [reply]
I also found the "**Dependent tasks that can be run in parallel**" section confusing. I think there are a few unstated assumptions.

The `"transit": { "dependsOn": ["^transit"] }` task, because it doesn't declare `inputs`, "include[s] all files in the package that are tracked by Git". So it will run whenever any file in the current package changes. And by depending on `^transit`, it will also run whenever any file in a package you depend on changes. 

Since there's no `transit` script in `package.json`, it won't actually _do_ anything by itself. But you can use `transit` as a task dependency to make a task like `check-types` run whenever any file in the current package or a package it depends on changes.

As for [motss](https://github.com/motss)'s examples, I'm not sure whether they intentionally used `build` instead of `lint` in the docs. Typically `build` has to depend on `^build`, so you can't parallelize it. But as for `lint`, like in the docs, I think this is the effect:

```
# Example 1 

## lint depends on ^lint
c#lint
       b#lint
              a#lint

## lint depends on transit, and transit depends on ^transit
c#transit
 b#transit
  a#transit
   a#lint
  b#lint
 c#lint
```

```
# Example 2

## lint depends on ^lint
c#lint
       b#lint
       a#lint

## lint depends on transit, and transit depends on ^transit
c#transit
 a#transit
 b#transit
  a#lint
  b#lint
 c#lint
```

I used the indentation to show how `transit` runs quickly and `lint` runs slowly.
